### DISCUSSION TITLE
undefined ref: cluster.skipfinalsnapshot.value when trivy is loading built-in policies

### DISCUSSION THREAD

**Discussion by rvillane** [aquasecurity/trivy#5258] [Bugs]
### Description

I just upgraded to trivy v0.45.1 and I'm getting an error when running the following command:

`trivy config --severity MEDIUM,HIGH,CRITICAL --exit-code 1 .`

### Desired Behavior

Run trivy scans without the issue loading the built-in policies

### Actual Behavior

trivy scan is failing with the following error message:

```
023-09-27T10:18:52.882Z	[34mINFO[0m	Downloading the built-in policies...
45.23 KiB / 45.23 KiB [-----------------------------------------------------------] 100.00% ? p/s 0s2023-09-27T10:18:53.827Z	[31mFATAL[0m	filesystem scan error: scan error: scan failed: failed analysis: failed to call hooks: post handler error: misconfiguration scan error: scan config error: 4 errors occurred:
home/jenkins/.cache/trivy/policy/content/policies/cloud/policies/aws/rds/disable_cluster_skip_final_snapshot.rego:26: rego_type_error: undefined ref: cluster.skipfinalsnapshot.value
	cluster.skipfinalsnapshot.value
	        ^
	        have: "skipfinalsnapshot"
	        want (one of): ["backupretentionperioddays" "encryption" "engine" "instances" "performanceinsights" "publicaccess" "replicationsourcearn"]
home/jenkins/.cache/trivy/policy/content/policies/cloud/policies/aws/rds/disable_cluster_skip_final_snapshot.rego:27: rego_type_error: undefined ref: cluster.skipfinalsnapshot
	cluster.skipfinalsnapshot
	        ^
	        have: "skipfinalsnapshot"
	        want (one of): ["backupretentionperioddays" "encryption" "engine" "instances" "performanceinsights" "publicaccess" "replicationsourcearn"]
home/jenkins/.cache/trivy/policy/content/policies/cloud/policies/aws/rds/enable_cluster_deletion_protection.rego:26: rego_type_error: undefined ref: cluster.deletionprotection.value
	cluster.deletionprotection.value
	        ^
	        have: "deletionprotection"
	        want (one of): ["backupretentionperioddays" "encryption" "engine" "instances" "performanceinsights" "publicaccess" "replicationsourcearn"]
home/jenkins/.cache/trivy/policy/content/policies/cloud/policies/aws/rds/enable_cluster_deletion_protection.rego:27: rego_type_error: undefined ref: cluster.deletionprotection
	cluster.deletionprotection
	        ^
	        have: "deletionprotection"
	        want (one of): ["backupretentionperioddays" "encryption" "engine" "instances" "performanceinsights" "publicaccess" "replicationsourcearn"]

```

### Reproduction Steps

```bash
1. Install trivy v0.45.1 in a container using FROM gcr.io/google.com/cloudsdktool/google-cloud-cli:slim as base image
2. trivy config --severity MEDIUM,HIGH,CRITICAL --exit-code 1 .
```

### Target

Filesystem

### Scanner

Misconfiguration

### Output Format

None

### Mode

Standalone

### Debug Output

```bash
023-09-27T10:18:52.882Z	[34mINFO[0m	Downloading the built-in policies...
45.23 KiB / 45.23 KiB [-----------------------------------------------------------] 100.00% ? p/s 0s2023-09-27T10:18:53.827Z	[31mFATAL[0m	filesystem scan error: scan error: scan failed: failed analysis: failed to call hooks: post handler error: misconfiguration scan error: scan config error: 4 errors occurred:
home/jenkins/.cache/trivy/policy/content/policies/cloud/policies/aws/rds/disable_cluster_skip_final_snapshot.rego:26: rego_type_error: undefined ref: cluster.skipfinalsnapshot.value
	cluster.skipfinalsnapshot.value
	        ^
	        have: "skipfinalsnapshot"
	        want (one of): ["backupretentionperioddays" "encryption" "engine" "instances" "performanceinsights" "publicaccess" "replicationsourcearn"]
home/jenkins/.cache/trivy/policy/content/policies/cloud/policies/aws/rds/disable_cluster_skip_final_snapshot.rego:27: rego_type_error: undefined ref: cluster.skipfinalsnapshot
	cluster.skipfinalsnapshot
	        ^
	        have: "skipfinalsnapshot"
	        want (one of): ["backupretentionperioddays" "encryption" "engine" "instances" "performanceinsights" "publicaccess" "replicationsourcearn"]
home/jenkins/.cache/trivy/policy/content/policies/cloud/policies/aws/rds/enable_cluster_deletion_protection.rego:26: rego_type_error: undefined ref: cluster.deletionprotection.value
	cluster.deletionprotection.value
	        ^
	        have: "deletionprotection"
	        want (one of): ["backupretentionperioddays" "encryption" "engine" "instances" "performanceinsights" "publicaccess" "replicationsourcearn"]
home/jenkins/.cache/trivy/policy/content/policies/cloud/policies/aws/rds/enable_cluster_deletion_protection.rego:27: rego_type_error: undefined ref: cluster.deletionprotection
	cluster.deletionprotection
	        ^
	        have: "deletionprotection"
	        want (one of): ["backupretentionperioddays" "encryption" "engine" "instances" "performanceinsights" "publicaccess" "replicationsourcearn"]

```
```


### Operating System

OCI container using ubuntu 

### Version

```bash
Version: 0.45.1
```

### Checklist

- [ ] Run `trivy image --reset`
- [X] Read [the troubleshooting](https://aquasecurity.github.io/trivy/latest/docs/references/troubleshooting/)

**Comment by knqyf263** [comment]
@simar7 @nikpivkin Any ideas?

**Comment by rvillane** [comment]
For now `--skip-policy-update` flag is helping to bypass this issue as it prevents to update the OPA policies that seems to be were the root cause is.

**Comment by simar7** [comment]
Hi we'll take a look at it. Having said that the rest of the policies should've scanned as correctly as we added the support to skip those policies that have issues. Was it not the case?

**Reply by rvillane** [reply]
well, it was until I added the `--skip-policy-update` flag that I was able to run a `trivy config` scan, otherwise it fails with the issue initially reported above

**Comment by simar7** [comment]
@rvillane can you please run the following and share the output?

```shell
$ trivy version
Version: 0.45.1
Vulnerability DB:
  Version: 2
  UpdatedAt: 2023-09-27 18:18:13.992102004 +0000 UTC
  NextUpdate: 2023-09-28 00:18:13.992101504 +0000 UTC
  DownloadedAt: 2023-09-27 22:49:01.692913 +0000 UTC
Policy Bundle:
  Digest: sha256:e67eca244bb8cd7783e0a7aee30125a2458804a25e0feb18b77f1fb6d717ac67
  DownloadedAt: 2023-09-27 02:41:00.114485 +0000 UTC
```

We've added support for handling newer policies that might not be supported by older versions of Trivy starting at Trivy version v0.42.0 and newer.

**Reply by rvillane** [reply]
when I run trivy version from the container, the output is abbreviated to:

```
Version: 0.45.1
```

When I run it locally on my Mac, I get an extended output:

```
$ trivy version
Version: 0.45.1
Vulnerability DB:
  Version: 2
  UpdatedAt: 2022-04-30 18:06:13.026857622 +0000 UTC
  NextUpdate: 2022-05-01 00:06:13.026857122 +0000 UTC
  DownloadedAt: 2022-04-30 18:18:40.177703 +0000 UTC
Policy Bundle:
  Digest: sha256:24b38cdf646f0e5becf55a709ae9a3c4e819a348c28990cec0b6aabe4637d8b1
  DownloadedAt: 2023-09-27 02:54:44.149776 +0000 UTC
```

To install trivy on the container image, I run the following instructions in the Dockerfile:

```
# install trivy
RUN wget -qO - https://aquasecurity.github.io/trivy-repo/deb/public.key | apt-key add - && \
    echo "deb https://aquasecurity.github.io/trivy-repo/deb $(lsb_release -sc) main" | tee -a /etc/apt/sources.list.d/trivy.list && \
    apt-get update && \
    apt-get install trivy
```

**Reply by simar7** [reply]
I tried reproducing this issue locally by having a container with the base image as `gcr.io/google.com/cloudsdktool/google-cloud-cli:slim` but was unable to. I have the same bundle SHA and Trivy version as you so that rules those two out. 

See the output here https://gist.github.com/simar7/70d5c3488febf454681e620b4721243a 

Would it be possible for you to share your complete dockerfile here so I can try to run that and see what's going on?

**Reply by rvillane** [reply]
sure, enclosed is the file with extension txt to allow github to upload the file
[Dockerfile.txt](https://github.com/aquasecurity/trivy/files/12746049/Dockerfile.txt)

**Reply by simar7** [reply]
thanks for providing that but I'm still unable to see the failure. You can see my output and how I run Trivy within the Dockerfile you shared here https://gist.github.com/simar7/43107d52a18ecb29be9315af205766d9

The result is a failure as Trivy successfully finds misconfigurations in the image and exits with a non zero exit code, as expected.

**Reply by rvillane** [reply]
I see you are getting the same abbreviated version output, however, I'm not clear why you are not getting the same error, perhaps you can just generate the image and then run the container instead of running the trivy command as part of the Dockerfile instruction ?

**Reply by simar7** [reply]
Hmm yeah it is very strange indeed. I just tried running the built image as you suggested but still no luck.

```shell
 docker run -it --rm docker.io/library/5258:latest config .
2023-09-29T21:22:58.793Z        INFO    Misconfiguration scanning is enabled
2023-09-29T21:22:58.793Z        INFO    Need to update the built-in policies
2023-09-29T21:22:58.793Z        INFO    Downloading the built-in policies...
45.23 KiB / 45.23 KiB [---------------------------------------------------------] 100.00% ? p/s 0s
2023-09-29T21:23:19.339Z        INFO    Detected config files: 30

usr/lib/google-cloud-sdk/platform/ext-runtime/go/data/Dockerfile (dockerfile)

Tests: 26 (SUCCESSES: 23, FAILURES: 3, EXCEPTIONS: 0)
Failures: 3 (UNKNOWN: 0, LOW: 1, MEDIUM: 1, HIGH: 1, CRITICAL: 0)

MEDIUM: Specify a tag in the 'FROM' statement for image 'gcr.io/google-appengine/golang'
══════════════════════════════════════════════════════════════════════════════════════════════════
When using a 'FROM' statement you should use a specific tag to avoid uncontrolled behavior when the image is updated.

<snip>
```

One thing I've noticed based on the SHA, is you have a different policy bundle than what version 0.45.1 ships with. Could you try to run `trivy image --reset` and run the scan again? This will remove any previous stale policy bundle artifacts and download the bundle again.

**Reply by rvillane** [reply]
ok, that was interesting, after running `trivy image --reset` and then running `trivy version`, now I'm getting a different output:

```
trivy version
Version: 0.45.1
Policy Bundle:
  Digest: sha256:24b38cdf646f0e5becf55a709ae9a3c4e819a348c28990cec0b6aabe4637d8b1
  DownloadedAt: 2023-09-30 00:36:17.272279369 +0000 UTC
```

**Reply by simar7** [reply]
Are you still running into the same issue or it was resolved after the reset?

**Reply by rvillane** [reply]
yes, that seems to be the case, I'm not able to reproduce the issue after the reset.  In a containerized context, is this reset recommended only when building the container where trivy will be installed? or is it recommended to reset every `trivy config` scan ?

**Reply by simar7** [reply]
I will investigate this further next week. Will get back to you.

**Reply by rvillane** [reply]
Hi @simar7 , just by chance do you have any updates about when reset needs to happen in a containerized context ?

**Reply by simar7** [reply]
I tried but I don't have any update for you. I'm not entirely sure why you see this behavior. As for the reset, you shouldn't need to do so but it is a way to clear any old cached policies and Trivy DB if you ever run into issues with caching.

**Comment by girlpunk** [comment]
For reference, I'm getting the same error running trivy 0.45.1 in MegaLinter 7.4.0

**Reply by simar7** [reply]
Are you able to share logs or more info?

**Reply by girlpunk** [reply]
I can, though they're identical to the ones rvillane posted in their original message, including the same 4 errors on the same 2 files.

**Reply by alterEgo123** [reply]
I have the same error for trivy 0.41.0

**Reply by simar7** [reply]
As mentioned, this issue was fixed in 0.42.0 and above. Please update to a recent version.

**Reply by nvuillam** [reply]
FYI MegaLinter 7.5.0 has been released and [embeds trivy v0.46.1](https://megalinter.io/latest/descriptors/repository_trivy/) :)

**Comment by nikpivkin** [comment]
@simar7 It seems to me that some old version of Trivy with new policies (which are cached) is being used, but definitely not v0.45.0. According to the error trace `scan error: scan failed: failed analysis: failed to call hooks: post handler error: misconfiguration scan error:`, it can be seen that it occurs when calling post handler (which refers to [this](https://github.com/nikpivkin/trivy/commit/bcc0218200dddcdcd9e3cef6de3339ae7c4440f5) PR). But now post analyzers are used. The logs also show that the cache is loaded from `home/jenkins/.cache`, but there is no such folder in the container above.

**Comment by evertonlsouza** [comment]
Hello,

I have the same problem, my scenario is as follows:

I'm using Azure Devops for CI/CD, I'm using the trivy@1 task, in the _Trivy Version_ field there is the default value "_latest_", I'm scanning my git path, until last week it was working.

steps:
- task: AquaSecurityOfficial.trivy-official.custom-build-release-task.trivy@1
  displayName: 'Echo trivy last git'
  Appetizer:
    docker: false
    path: '$(Build.Repository.LocalPath)'
    severities: 'CRITICAL, HIGH, MEDIUM'
    ignoreUnfixed: true
    exit code: 0

Error:
FATAL	filesystem scan error: scan error: scan failed: failed analysis: failed to call hooks: post handler error: misconfiguration scan error: scan config error: 4 errors occurred:
policies/cloud/policies/aws/rds/disable_cluster_skip_final_snapshot.rego:26: rego_type_error: undefined ref: cluster.skipfinalsnapshot.value

Observation:
I'm using a Microsoft-hosted agent build, with each pipe started an empty agent is created and Trivy is installed according to the task specification

**Reply by evertonlsouza** [reply]
I did a test using Trivy version v0.45.0 and it worked, it seems the problem is with the latest version.

- task: AquaSecurityOfficial.trivy-official.custom-build-release-task.trivy@1
  displayName: 'Echo trivy latest git'
  inputs:
    version: v0.45.0
    docker: false
    path: '$(Build.Repository.LocalPath)'
    severities: 'CRITICAL,HIGH,MEDIUM'
    ignoreUnfixed: true
    exitCode: 0

**Comment by simar7** [comment]
@nikpivkin @evertonlsouza my guess is that this is why it's breaking https://github.com/aquasecurity/trivy-azure-pipelines-task/blob/main/trivy-task/index.ts#L7 - it uses a very old version of Trivy. 

I've made a PR to handle that https://github.com/aquasecurity/trivy-azure-pipelines-task/pull/49

**Comment by vintury** [comment]
Hello.
Any news? I have same issue.

**Reply by evertonlsouza** [reply]
They recently merged the Azure DevOps task (https://github.com/aquasecurity/trivy-azure-pipelines-task/pull/49) but when testing my pipeline the error remains, it only works if I put version v0.45.0

Error:
FATAL filesystem scan error: scan error: scan failed: failed analysis: failed to call hooks: post handler error: misconfiguration scan error: scan config error: 4 errors occurred:
policies/cloud/policies/aws/rds/disable_cluster_skip_final_snapshot.rego:26: rego_type_error: undefined ref: cluster.skipfinalsnapshot.value
cluster.skipfinalsnapshot.value

**Reply by simar7** [reply]
> They recently merged the Azure DevOps task (https://github.com/aquasecurity/trivy-azure-pipelines-task/pull/49) but when testing my pipeline the error remains, it only works if I put version v0.45.0
> 
> 
> 
> Error:
> 
> FATAL filesystem scan error: scan error: scan failed: failed analysis: failed to call hooks: post handler error: misconfiguration scan error: scan config error: 4 errors occurred:
> 
> policies/cloud/policies/aws/rds/disable_cluster_skip_final_snapshot.rego:26: rego_type_error: undefined ref: cluster.skipfinalsnapshot.value
> 
> cluster.skipfinalsnapshot.value

Are you using [this](https://github.com/aquasecurity/trivy-azure-pipelines-task/releases/tag/v1.4.2) released version which includes the fix?

**Reply by evertonlsouza** [reply]
In the trivy version field I used "latest" and it gave an error, using version "v0.45.0" it works correctly

**Reply by simar7** [reply]
Are you talking about this? https://github.com/aquasecurity/trivy-azure-pipelines-task/blob/main/trivy-task/index.ts#L7 Isn't that set to 0.45.1? Or can you share your pipeline so we can take a look at what's going on?

**Reply by evertonlsouza** [reply]
My pipeline that doesn't work:

```
trigger:
- main
 
pool:
  name: my-agent

steps:
- task: Docker@2
  displayName: login
  inputs:
    containerRegistry: my-acr
    command: login

- task: AquaSecurityOfficial.trivy-official.custom-build-release-task.trivy@1
  displayName: 'Run Trivy on the Repos'
  inputs:
    docker: false
    path: '$(System.DefaultWorkingDirectory)'
    severities: 'CRITICAL,HIGH'
    exitCode: 0

- task: AquaSecurityOfficial.trivy-official.custom-build-release-task.trivy@1
  displayName: 'Run Trivy in ACR'
  inputs:
    docker: false
    loginDockerConfig: true
    image: 'my-image'
    severities: 'HIGH,CRITICAL'
    exitCode: 0
  

```
 Error:
FATAL	filesystem scan error: scan error: scan failed: failed analysis: failed to call hooks: post handler error: misconfiguration scan error: scan config error: 4 errors occurred:
policies/cloud/policies/aws/rds/disable_cluster_skip_final_snapshot.rego:26: rego_type_error: undefined ref: cluster.skipfinalsnapshot.value
	cluster.skipfinalsnapshot.value



If I include the "version: v0.45.0", it works:

```
trigger:
- main
 
pool:
  name: my-agent

steps:
- task: Docker@2
  displayName: login
  inputs:
    containerRegistry: my-acr
    command: login

- task: AquaSecurityOfficial.trivy-official.custom-build-release-task.trivy@1
  displayName: 'Run Trivy on the Repos'
  inputs:
    version: v0.45.0
    docker: false
    path: '$(System.DefaultWorkingDirectory)'
    severities: 'CRITICAL,HIGH'
    exitCode: 0

- task: AquaSecurityOfficial.trivy-official.custom-build-release-task.trivy@1
  displayName: 'Run Trivy in ACR'
  inputs:
    version: v0.45.0
    docker: false
    loginDockerConfig: true
    image: 'my-image'
    severities: 'HIGH,CRITICAL'
    exitCode: 0
```

**Reply by simar7** [reply]
Thanks for sharing but something doesn't add up. I took a look at the code (sorry I'm not a maintainer so trying my best guess here) and it seems that if no version is provided, we use the latest. https://github.com/aquasecurity/trivy-azure-pipelines-task/blob/main/trivy-task/index.ts#L197-L199

Which in this case is set to `v0.45.1` https://github.com/aquasecurity/trivy-azure-pipelines-task/blob/main/trivy-task/index.ts#L7

This issue was well fixed in `v0.42.0` and your manually specifying trivy version even at `v0.45.0` (lower than the one action defaults to on latest) confirms that. Which is why I don't understand why the runner picks up an older version of trivy at runtime within the pipeline. 

The best I can suggest is to debug within your pipeline to find out if anything is getting cached by azure pipelines. You can see the previous discussion here https://github.com/aquasecurity/trivy/discussions/5258#discussioncomment-7149303 Otherwise the only workaround is for now would be to manually specify the trivy version in the pipeline task like you did, to explicitly specify the version to the runner.

**Reply by simar7** [reply]
@evertonlsouza I think this might be a bug in the azure pipeline task after all, see this issue https://github.com/aquasecurity/trivy-azure-pipelines-task/issues/45 for details.

**Reply by evertonlsouza** [reply]
From what I understand the problem is when the version is not informed, I did some tests using the version v0.45.0 and v0.48.3 and it worked, I did the test using two types of pool agent, Self-hosted and MS-hosted, if I inform the version the pipe runs successfully, otherwise it breaks.

I believe that at this point it is best to inform the version
